lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -1640,17 +1654,118 @@ namespace Js
     }
 #endif
 
-    Var IntlEngineInterfaceExtensionObject::EntryIntl_CompareString(RecyclableObject* function, CallInfo callInfo, ...)
+#ifdef INTL_ICU
+    static int CompareStringICU(_In_opt_count_(langtagLen) const char16 *langtag, _In_ charcount_t langtagLen, _In_z_ const char16 *left, _In_ charcount_t cchLeft, _In_z_ const char16 *right, _In_ charcount_t cchRight,
+        _In_ CollatorSensitivity sensitivity, _In_ bool ignorePunctuation, _In_ bool numeric, _In_ CollatorCaseFirst caseFirst, _Out_ HRESULT *hr)
     {
-        EngineInterfaceObject_CommonFunctionProlog(function, callInfo);
+        ASSERT_ENUM(CollatorSensitivity, sensitivity);
+        ASSERT_ENUM(CollatorCaseFirst, caseFirst);
+        Assert(left != nullptr && right != nullptr && hr != nullptr);
+        int ret = 0;
+        *hr = E_INVALIDARG;
+        UErrorCode error = U_ZERO_ERROR;
 
-        if (args.Info.Count < 3 || !JavascriptString::Is(args.Values[1]) || !JavascriptString::Is(args.Values[2]))
+        char localeID[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int32_t length = 0;
+        if (langtag == nullptr)
         {
-            JavascriptError::MapAndThrowError(scriptContext, E_INVALIDARG);
+            length = uloc_getName(nullptr, localeID, _countof(localeID), &error);
         }
+        else
+        {
+            char langtag8[ULOC_FULLNAME_CAPACITY] = { 0 };
+            AssertOrFailFast(utf8::WideStringToNarrowNoAlloc(langtag, langtagLen, langtag8, _countof(langtag8)) == S_OK);
+            uloc_forLanguageTag(langtag8, localeID, _countof(localeID), &length, &error);
+        }
+        ICU_ASSERT(error, length > 0);
+
+        ScopedUCollator collator(ucol_open(localeID, &error));
+        ICU_ASSERT(error, true);
 
+        if (sensitivity == CollatorSensitivity::Base)
+        {
+            ucol_setStrength(collator, UCOL_PRIMARY);
+        }
+        else if (sensitivity == CollatorSensitivity::Accent)
+        {
+            ucol_setStrength(collator, UCOL_SECONDARY);
+        }
+        else if (sensitivity == CollatorSensitivity::Case)
+        {
+            // see "description" for the caseLevel default option: http://userguide.icu-project.org/collation/customization
+            ucol_setStrength(collator, UCOL_PRIMARY);
+            ucol_setAttribute(collator, UCOL_CASE_LEVEL, UCOL_ON, &error);
+            ICU_ASSERT(error, true);
+        }
+        else if (sensitivity == CollatorSensitivity::Variant)
+        {
+            ucol_setStrength(collator, UCOL_TERTIARY);
+        }
+        else
+        {
+            AssertMsg(false, "sensitivity is not one of the CollatorSensitivity values");
```

Comment:
>AssertMsg [](start = 12, length = 9)

What happens in retail if we fall-through here? I don't see any reason this shouldn't failfast.

Reply:
It was this way before, but that can change.